### PR TITLE
[loxone] Support for Intelligent Room Controller #6433

### LINKED ISSUE #6433: [loxone] Support for Heating
Hi there,
would it be possible to add the Intelligent Heat Controller?

Thanks in advance

Mees Maarten

### PR DESCRIPTION
**Pull request by ppieczul**
Added support for Intelligent Room Controller V1 

Signed-off-by: Pawel Pieczul <pieczul@gmail.com>

Fixes #6433

### CODE REVIEW

**Comment by wborn** [comment]
Thanks for adding support so quickly @ppieczul!
Perhaps you also have a link to a JAR containing the changes so @MaartenMees can help testing?

**Comment by ppieczul** [comment]
Yes, this is already happening, I am in a separate conversation with
Maarten.

**Comment by ppieczul** [comment]
@worn what should I do to get this merged? I can see there is 2.5 branch and master has some different organisation of files. Should I get this on 2.5 branch or master? And when?

**Comment by wborn** [comment]
See https://github.com/openhab/openhab-addons/issues/6702. You can switch the PR by editing it and changing the base from master to 2.5.x. After that you'll probably have to rebase your branch on 2.5.x and force push the result.

**Comment by ppieczul** [comment]
@wborn, thanks I did as you advised. I think this feature is complete.  @MaartenMees can you please confirm or add more feedback. Thanks.

**Comment by martinvw** [comment]
@ppieczul do you want us to wait for @MaartenMees or can we just continue?

**Comment by ppieczul** [comment]
Please go ahead and review/merge. I asked Marteen for feedback and he
replied he would check it out but probably had no time.

**Comment by MaartenMees** [comment]
Sorry guys, I am swamped at the moment. Real sorry..
________________________________

**Comment by martinvw** [comment]
@openhab/add-ons-maintainers anyone else can take a quick look? Thanks!

File: bundles/org.openhab.binding.loxone/README.md [hunk 1]
```diff
@@ -99,6 +99,14 @@ In case a websocket connection to the Miniserver remains active for the whole du
 
 A method to enable unrestricted security policy depends on the JRE version and vendor, some examples can be found [here](https://www.petefreitag.com/item/844.cfm) and [here](https://stackoverflow.com/questions/41580489/how-to-install-unlimited-strength-jurisdiction-policy-files).
 
+### Upgrading Miniserver's Firmware
+
+After you upgrade Miniserver's firmware to a new version, you may experience user authentication error, when binding is connecting to the Miniserver. This is due to the Miniserver forgetting the authentication tokens after the upgrade.
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.loxone/README.md]: Please put each sentence on a new line.

File: bundles/org.openhab.binding.loxone/README.md [hunk 2]
```diff
@@ -315,3 +347,4 @@ OFF=Closed
 ON=Open
 -=Unknown
 ```
+222
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.loxone/README.md]: `222` accidental key hits?

File: bundles/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/controls/LxControl.java [hunk 3]
```diff
@@ -705,6 +717,20 @@ void removeAllChannels() {
...
 
+    /**
+     * Adds a new state to the control.
+     *
+     * @param name name of the state
+     * @param uuid UUID of the state - can be null
+     */
+    private void addState(String name, String uuid) {
+        if (uuid == null) {
+            logger.warn("Unexpected null UUID in state {} inside state array in control {}", name, this.name);
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/controls/LxControl.java]: I'm not sure what this warn message intend is? If the user would see this message what should the user do? If the user can't do anything useful here it should either be debug log or the message should give a clear message what the consequence is when this is null and possible how to (possible) fix it.

File: bundles/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/controls/LxControlIRCDayTimer.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        }
+    }
+
+    private static final String STATE_CURRENT_TEMPERATURE_INDEX = "value";
+    private static final String STATE_CURRENT_MODE = "mode";
+    private static final String STATE_MODE_LIST = "modelist";
+
+    private Map<Integer, String> modesList = new HashMap<>();
+
+    private static Pattern pattern;
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/controls/LxControlIRCDayTimer.java]:
```suggestion
    private static final Pattern PATTERN = Pattern.compile("(\\d+):mode=([\\d-]+);name=\"(.+)\"");
```

File: bundles/org.openhab.binding.loxone/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 5]
```diff
@@ -206,7 +206,71 @@
 		<state readOnly="true"></state>
 	</channel-type>
 
-	<channel-type id="iRoomV2ActiveModeTypeId">
+	<channel-type id="iRoomModeTypeId">
+		<item-type>Number</item-type>
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.loxone/src/main/resources/ESH-INF/thing/thing-types.xml]: Officially channels that have options should be of item-type String: https://www.openhab.org/docs/developer/bindings/thing-xml.html#state-description

File: bundles/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/types/LxIRCTemperature.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                return t;
+            }
+        }
+        return UNKNOWN;
+    }
+
+    public String getLabel() {
+        switch (index) {
+            case 0:
+                return "Economy";
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/types/LxIRCTemperature.java]: This is all static so it should be a parameter of the enum: `ECONOMY(0, "Economy")`

File: bundles/org.openhab.binding.loxone/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 5 above)
Comment (ppieczul) [reply bundles/org.openhab.binding.loxone/src/main/resources/ESH-INF/thing/thing-types.xml]: @Hilbrand Do you know some background why this is advised to be string? I am thinking what should be the value in such case, just the number as a string or the label transformed into some short text form, so the user can use it when sending a command to the channel?

Comment (Hilbrand) [reply bundles/org.openhab.binding.loxone/src/main/resources/ESH-INF/thing/thing-types.xml:210]: I did some research on where this came from and my advice here seems misguided. It is allowed to have a number as type and the documentation is wrong on this. Although I would advice to use meaningful names when numbers don't have any reference for the user. With semi random numbers it's easy for the user to pick the wrong number or have to look up the number in the readme to see what it means. So you can still keep the Number type if you want to.

**Comment by kaikreuzer** [comment]
Closing this PR as it is not in a mergeable state. Please follow the instructions above to port this PR to the main branch for openHAB 3.

**Pull request closed without merging**